Creation PackageURL objects for invalid PURLs / Possible improper encoding of colons (":") in PURL fields

It is possible to create PackageURL objects that contain invalid fields, specifically by using the `PackageURL` kwarg constructor and passing in values that contain colons.

Simple example:
```
>>> from packageurl import PackageURL
>>> p = PackageURL(type="generic", name="Foo: <Bar>", version="1.2.3")
>>> p.to_string()
'pkg:generic/Foo:%20%3CBar%3E@1.2.3'
>>> PackageURL.from_string(p.to_string())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vossn/finitestate/finite-state-sip/venv/lib/python3.10/site-packages/packageurl/__init__.py", line 514, in from_string
    raise ValueError(msg)
ValueError: Invalid purl 'pkg:generic/Foo:%20%3CBar%3E@1.2.3' cannot contain a "user:pass@host:port" URL Authority component: ''.
```

On closer inspection, it looks like the problem might be that colons (`:`) are not being percent-encoded correctly? I would expect the colon in the name to be encoded to `%3A`, but it looks like it is being left as a literal `:` in the `to_string()` function:
```
>>> p = PackageURL(type="generic", name="Foo: <Bar>", version="1.2.3")
>>> p
PackageURL(type='generic', namespace=None, name='Foo: <Bar>', version='1.2.3', qualifiers={}, subpath=None)
>>> p.to_string()
'pkg:generic/Foo:%20%3CBar%3E@1.2.3'
```

I'm not sure I'm interpreting the [PURL spec](https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst) correctly with regards to the treatment of colon characters, but on the surface it sounds like any colon appearing within an individual field value should simply be percent-encoded if the field itself calls for it.

OK, so let's talk about the standard, from here: https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst

Namespace:
> Each namespace segment must be a percent-encoded string

Name:
> A name must be a percent-encoded string

Version:
> A version must be a percent-encoded string

Qualifiers:
> A key must NOT be percent-encoded

> A value must be a percent-encoded string

Subpath:
> Each subpath segment must be a percent-encoded string

So, it would appear there are incorrect tests. For example:
```
E       AssertionError: assert 'pkg:docker/c...ry_url=gcr.io' == 'pkg:docker/c...ry_url=gcr.io'
E
E         - pkg:docker/customer/dockerimage@sha256%3A244fd47e07d1004f0aed9c?repository_url=gcr.io
E         ?                                       ^^^
E         + pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io
E         ?
```
Hmm...does this mean the test data is wrong?
```
  {
    "description": "docker uses qualifiers and hash image id as versions",
    "purl": "pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io",
    "canonical_purl": "pkg:docker/customer/dockerimage@sha256:244fd47e07d1004f0aed9c?repository_url=gcr.io",
    "type": "docker",
    "namespace": "customer",
    "name": "dockerimage",
    "version": "sha256:244fd47e07d1004f0aed9c",
    "qualifiers": {"repository_url": "gcr.io"},
    "subpath": null,
    "is_invalid": false
  },
```
If the version is supposed to be percent-encoded, then a version of `sha256:244fd47e07d1004f0aed9c` should end up as `sha256%3A244fd47e07d1004f0aed9c`

How do we go forward with this?

This is also incorrect:
```
E       AssertionError: 'pkg:[42 chars]https://github.com/z3APA3A/3proxy/releases/dow[26 chars].zip' != 'pkg:[42 chars]https%3A//github.com/z3APA3A/3proxy/releases/d[28 chars].zip'
E       - pkg:github/z3apa3a/3proxy@0.8.11?download_url=https://github.com/z3APA3A/3proxy/releases/download/0.8.11/3proxy-0.8.11.zip
E       ?                                                    ^
E       + pkg:github/z3apa3a/3proxy@0.8.11?download_url=https%3A//github.com/z3APA3A/3proxy/releases/download/0.8.11/3proxy-0.8.11.zip
```
The spec says the value of a qualifier *must be* percent-encoded. Thus, the `:` and `/` characters in the URL should be percent encoded.

Where do I go from here?

--- a/src/packageurl/__init__.py
+++ b/src/packageurl/__init__.py
@@ -77,11 +77,13 @@ def unquote(s: AnyStr) -> str:
 
 
 @overload
-def get_quoter(encode: bool = True) -> Callable[[AnyStr], str]: ...
+def get_quoter(encode: bool = True) -> Callable[[AnyStr], str]:
+    ...
 
 
 @overload
-def get_quoter(encode: None) -> Callable[[str], str]: ...
+def get_quoter(encode: None) -> Callable[[str], str]:
+    ...
 
 
 def get_quoter(encode: bool | None = True) -> Callable[[AnyStr], str] | Callable[[str], str]:
@@ -151,19 +153,22 @@ def normalize_version(version: AnyStr | None, encode: bool | None = True) -> str
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: Literal[True] = ...
-) -> str | None: ...
+) -> str | None:
+    ...
 
 
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: Literal[False] | None
-) -> dict[str, str]: ...
+) -> dict[str, str]:
+    ...
 
 
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: bool | None = ...
-) -> str | dict[str, str] | None: ...
+) -> str | dict[str, str] | None:
+    ...
 
 
 def normalize_qualifiers(
@@ -251,7 +256,8 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: Literal[True] = ...,
-) -> tuple[str, str | None, str, str | None, str | None, str | None]: ...
+) -> tuple[str, str | None, str, str | None, str | None, str | None]:
+    ...
 
 
 @overload
@@ -263,7 +269,8 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: Literal[False] | None,
-) -> tuple[str, str | None, str, str | None, dict[str, str], str | None]: ...
+) -> tuple[str, str | None, str, str | None, dict[str, str], str | None]:
+    ...
 
 
 @overload
@@ -275,7 +282,8 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: bool | None = ...,
-) -> tuple[str, str | None, str, str | None, str | dict[str, str] | None, str | None]: ...
+) -> tuple[str, str | None, str, str | None, str | dict[str, str] | None, str | None]:
+    ...
 
 
 def normalize(
@@ -459,12 +467,17 @@ class PackageURL(
             url=remainder, scheme="", allow_fragments=True
         )
 
-        if scheme or authority:
-            msg = (
-                f'Invalid purl {purl!r} cannot contain a "user:pass@host:port" '
-                f"URL Authority component: {authority!r}."
-            )
-            raise ValueError(msg)
+        # The spec (seems) to allow colons in the name and namespace.
+        # urllib.urlsplit splits on : considers them parts of scheme
+        # and authority.
+        # Other libraries do not care about this.
+        # See https://github.com/package-url/packageurl-python/issues/152#issuecomment-2637692538
+        # We do + ":" + to put the colon back that urlsplit removed.
+        if authority:
+            path = authority + ":" + path
+
+        if scheme:
+            path = scheme + ":" + path
 
         path = path.lstrip("/")
 

--- a/src/packageurl/__init__.py
+++ b/src/packageurl/__init__.py
@@ -77,13 +77,11 @@ def unquote(s: AnyStr) -> str:
 
 
 @overload
-def get_quoter(encode: bool = True) -> Callable[[AnyStr], str]:
-    ...
+def get_quoter(encode: bool = True) -> Callable[[AnyStr], str]: ...
 
 
 @overload
-def get_quoter(encode: None) -> Callable[[str], str]:
-    ...
+def get_quoter(encode: None) -> Callable[[str], str]: ...
 
 
 def get_quoter(encode: bool | None = True) -> Callable[[AnyStr], str] | Callable[[str], str]:
@@ -153,22 +151,19 @@ def normalize_version(version: AnyStr | None, encode: bool | None = True) -> str
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: Literal[True] = ...
-) -> str | None:
-    ...
+) -> str | None: ...
 
 
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: Literal[False] | None
-) -> dict[str, str]:
-    ...
+) -> dict[str, str]: ...
 
 
 @overload
 def normalize_qualifiers(
     qualifiers: AnyStr | dict[str, str] | None, encode: bool | None = ...
-) -> str | dict[str, str] | None:
-    ...
+) -> str | dict[str, str] | None: ...
 
 
 def normalize_qualifiers(
@@ -256,8 +251,7 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: Literal[True] = ...,
-) -> tuple[str, str | None, str, str | None, str | None, str | None]:
-    ...
+) -> tuple[str, str | None, str, str | None, str | None, str | None]: ...
 
 
 @overload
@@ -269,8 +263,7 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: Literal[False] | None,
-) -> tuple[str, str | None, str, str | None, dict[str, str], str | None]:
-    ...
+) -> tuple[str, str | None, str, str | None, dict[str, str], str | None]: ...
 
 
 @overload
@@ -282,8 +275,7 @@ def normalize(
     qualifiers: AnyStr | dict[str, str] | None,
     subpath: AnyStr | None,
     encode: bool | None = ...,
-) -> tuple[str, str | None, str, str | None, str | dict[str, str] | None, str | None]:
-    ...
+) -> tuple[str, str | None, str, str | None, str | dict[str, str] | None, str | None]: ...
 
 
 def normalize(

--- a/tests/test_packageurl.py
+++ b/tests/test_packageurl.py
@@ -342,7 +342,7 @@ def test_colons_in_name_are_handled_correctly() -> None:
     assert p.name == "libiconv: character set conversion library"
     assert p.version == "1.9"
     assert p.qualifiers == {"package-id": "e11a609df352e292"}
-    assert p.subpath == None
+    assert p.subpath is None
 
     assert PackageURL.from_string(p.to_string()).to_string() == p.to_string()
 
@@ -357,7 +357,7 @@ def test_colons_in_namespace_are_handled_correctly() -> None:
     assert p.name == "libiconv: character set conversion library"
     assert p.version == "1.9"
     assert p.qualifiers == {"package-id": "e11a609df352e292"}
-    assert p.subpath == None
+    assert p.subpath is None
 
     assert PackageURL.from_string(p.to_string()).to_string() == p.to_string()
 

--- a/tests/test_packageurl.py
+++ b/tests/test_packageurl.py
@@ -330,3 +330,47 @@ class NormalizePurlTest(unittest.TestCase):
 def test_purl_is_hashable():
     s = {PackageURL(name="hashable", type="pypi")}
     assert len(s) == 1
+
+
+def test_colons_in_name_are_handled_correctly() -> None:
+    p = PackageURL.from_string(
+        "pkg:nuget/libiconv:%20character%20set%20conversion%20library@1.9?package-id=e11a609df352e292"
+    )
+
+    assert p.type == "nuget"
+    assert p.namespace is None
+    assert p.name == "libiconv: character set conversion library"
+    assert p.version == "1.9"
+    assert p.qualifiers == {"package-id": "e11a609df352e292"}
+    assert p.subpath == None
+
+    assert PackageURL.from_string(p.to_string()).to_string() == p.to_string()
+
+
+def test_colons_in_namespace_are_handled_correctly() -> None:
+    p = PackageURL.from_string(
+        "pkg:nuget/an:odd:space/libiconv:%20character%20set%20conversion%20library@1.9?package-id=e11a609df352e292"
+    )
+
+    assert p.type == "nuget"
+    assert p.namespace == "an:odd:space"
+    assert p.name == "libiconv: character set conversion library"
+    assert p.version == "1.9"
+    assert p.qualifiers == {"package-id": "e11a609df352e292"}
+    assert p.subpath == None
+
+    assert PackageURL.from_string(p.to_string()).to_string() == p.to_string()
+
+
+def test_encoding_stuff_with_colons_correctly() -> None:
+    p = PackageURL(
+        type="nuget",
+        namespace="an:odd:space",
+        name="libiconv: character set conversion library",
+        version="1.9",
+        qualifiers={"package-id": "e11a609df352e292"},
+    )
+    assert (
+        p.to_string()
+        == "pkg:nuget/an:odd:space/libiconv:%20character%20set%20conversion%20library@1.9?package-id=e11a609df352e292"
+    )

